stdlib/public/Concurrency/Task.cpp

```diff
@@ -21,20 +21,129 @@
 #include "TaskPrivate.h"
 
 using namespace swift;
+using FutureFragment = AsyncTask::FutureFragment;
+
+size_t FutureFragment::fragmentSize(const Metadata *resultType) {
+  return storageOffset(resultType) + resultType->vw_size();
+}
+
+void FutureFragment::destroy() {
+  switch (status.load()) {
+  case Status::Executing:
+    assert(false && "destroying a task that never completed");
+
+  case Status::Success:
+    resultType->vw_destroy(getStoragePtr());
+    break;
+
+  case Status::Error:
+    swift_unknownObjectRelease(getStoragePtr());
+    break;
+  }
+}
+
+FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask) {
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  auto currentStatus = fragment->status.load();
+  switch (currentStatus) {
+  case FutureFragment::Status::Error:
+  case FutureFragment::Status::Success:
+    // The task is done; we don't need to wait.
+    return currentStatus;
+
+  case FutureFragment::Status::Executing:
+    break;
+  }
+
+  // Put the waiting task at the beginning of the wait queue.
+  while (true) {
+    waitingTask->NextWaitingTask = fragment->waitQueue.load();
+    if (fragment->waitQueue.compare_exchange_strong(
+            waitingTask->NextWaitingTask, waitingTask)) {
+      // Escalate the priority of this task based on the priority
+      // of the waiting task.
+      swift_task_escalate(this, waitingTask->Flags.getPriority());
+      break;
+    }
+  }
+
+  return FutureFragment::Status::Executing;
+}
+
+void AsyncTask::completeFuture(AsyncContext *context) {
+  assert(isFuture());
+  auto fragment = futureFragment();
+  auto storagePtr = fragment->getStoragePtr();
+
+  // Check for an error.
+  if (unsigned errorOffset = fragment->errorOffset) {
+    // Find the error object in the context.
+    auto errorPtrPtr = reinterpret_cast<char *>(context) + errorOffset;
+    OpaqueValue *errorObject = *reinterpret_cast<OpaqueValue **>(errorPtrPtr);
+
+    // If there is an error, take it and we're done.
+    if (errorObject) {
+      *reinterpret_cast<OpaqueValue **>(storagePtr) = errorObject;
+      fragment->status = FutureFragment::Status::Error;
+      return;
+    }
+  }
+
+  // Take the success value.
+  auto resultPtr = reinterpret_cast<OpaqueValue *>(
+      reinterpret_cast<char *>(context) + fragment->resultOffset);
+  fragment->resultType->vw_initializeWithTake(storagePtr, resultPtr);
+  fragment->status = FutureFragment::Status::Success;
```

Comment:
The general ABI for indirect return values is that the address to fill would be passed as an argument, not that we'd make space for it in the initial context and that move out of that into the future.

I don't remember what convention we decided on for errors, but I doubt this it's this strange.  Should there be a flag in the task reporting whether it can throw?